libbpf-tools/ext4dist.c

```diff
@@ -0,0 +1,252 @@
+	.times = 99999999,
+};
+
+static volatile bool exiting;
+
+const char *argp_program_version = "ext4dist 0.1";
+const char *argp_program_bug_address =
+	"https://github.com/iovisor/bcc/tree/master/libbpf-tools";
+const char argp_program_doc[] =
+"Summarize ext4 operation latency.\n"
+"\n"
+"Usage: ext4dist [-h] [-T] [-m] [-p PID] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    ext4dist          # show operation latency as a histogram\n"
+"    ext4dist -p 181   # trace PID 181 only\n"
+"    ext4dist 1 10     # print 1 second summaries, 10 times\n"
+"    ext4dist -m 5     # 5s summaries, milliseconds\n";
+
+static const struct argp_option opts[] = {
+	{ "timestamp", 'T', NULL, 0, "Print timestamp" },
+	{ "milliseconds", 'm', NULL, 0, "Millisecond histogram" },
+	{ "pid", 'p', "PID", 0, "Process PID to trace" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case 'm':
+		env.milliseconds = true;
+		break;
+	case 'p':
+		errno = 0;
+		env.pid = strtol(arg, NULL, 10);
+		if (errno || env.pid <= 0) {
+			fprintf(stderr, "invalid PID: %s\n", arg);
+			argp_usage(state);
+		}
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static char *fop_names[] = {
+	[READ_ITER] = "read_iter",
+	[WRITE_ITER] = "write_iter",
+	[OPEN] = "open",
+	[FSYNC] = "fsync",
+};
+
+static struct hist zero;
+
+static int print_hists(struct ext4dist_bpf__bss *bss)
+{
+	const char *units = env.milliseconds ? "msecs" : "usecs";
+	enum ext4_fop_type type;
+
+	for (type = READ_ITER; type < __MAX_FOP_TYPE; type++) {
+		struct hist hist = bss->hists[type];
+
+		bss->hists[type] = zero;
+		if (!memcmp(&zero, &hist, sizeof(hist)))
+			continue;
+		printf("operation = '%s'\n", fop_names[type]);
+		print_log2_hist(hist.slots, MAX_SLOTS, units);
+		printf("\n");
+	}
+
+	return 0;
+}
+
+static bool should_fallback(void)
+{
+	if (is_kernel_module("ext4") && !access("/sys/kernel/btf/ext4", R_OK))
+		return true;
+	return false;
+}
+
+static void ext4dist__choose_progs(struct ext4dist_bpf *skel, bool fallback)
```

Comment:
maybe just inline this, this function doesn't serve much purpose beyond distracting from linearly following the code in main()

Reply:
Thanks for your suggestion, I will fix it

Reply:
Fixed, thanks!